From 19acc0c777479f267d9c82c90cc7717dd6bb7345 Mon Sep 17 00:00:00 2001
From: Bernardo Garza <bernardo.garza@hlioengineering.com>
Date: Wed, 20 Apr 2022 16:03:16 -0500
Subject: [PATCH] Updated the atmel driver to add support for the
 touchscreen-inverted-x and touchscreen-inverted-y parameters to invert the
 orientation of the touch for the corresponding axis.

---
 .../devicetree/bindings/input/atmel,maxtouch.txt    |  3 +++
 drivers/input/touchscreen/atmel_mxt_ts.c            | 13 ++++++++++++-
 2 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/Documentation/devicetree/bindings/input/atmel,maxtouch.txt b/Documentation/devicetree/bindings/input/atmel,maxtouch.txt
index c88919480..33bd256dd 100644
--- a/Documentation/devicetree/bindings/input/atmel,maxtouch.txt
+++ b/Documentation/devicetree/bindings/input/atmel,maxtouch.txt
@@ -31,6 +31,9 @@ Optional properties for main touchpad device:
 
 - reset-gpios: GPIO specifier for the touchscreen's reset pin (active low)
 
+- touchscreen-inverted-x: Invert the X axis on the touch driver
+- touchscreen-inverted-y: Invert the Y axis on the touch driver
+
 Example:
 
 	touch@4b {
diff --git a/drivers/input/touchscreen/atmel_mxt_ts.c b/drivers/input/touchscreen/atmel_mxt_ts.c
index 3699af93d..fdef9b7cb 100644
--- a/drivers/input/touchscreen/atmel_mxt_ts.c
+++ b/drivers/input/touchscreen/atmel_mxt_ts.c
@@ -417,6 +417,9 @@ struct mxt_data {
 	u8 crc_err_count;
 	bool is_resync_enabled;
 
+	bool touchscreen_inverted_x;
+	bool touchscreen_inverted_y;
+
 	/* Cached parameters from object table */
 	u16 T5_address;
 	u8 T5_msg_size;
@@ -1467,7 +1470,12 @@ static void mxt_proc_t100_message(struct mxt_data *data, u8 *message)
 	status = message[1];
 	x = get_unaligned_le16(&message[2]);
 	y = get_unaligned_le16(&message[4]);
-	
+
+	if (data->touchscreen_inverted_x)
+		x = data->max_x - x;
+	if (data->touchscreen_inverted_y)
+		y = data->max_y - y;
+
 	/* Get other auxdata[] bytes if present */
 	
 	if (status & MXT_T100_DETECT) {
@@ -5189,6 +5197,9 @@ static int mxt_parse_device_properties(struct mxt_data *data)
 
 	data->is_resync_enabled = device_property_read_bool(dev, "resync-enabled");
 
+	data->touchscreen_inverted_x = device_property_read_bool(dev, "touchscreen-inverted-x");
+	data->touchscreen_inverted_y = device_property_read_bool(dev, "touchscreen-inverted-y");
+
 	return 0;
 }
 
-- 
2.25.1

